presto-main/src/main/java/com/facebook/presto/split/BufferingSplitSource.java

```diff
@@ -44,22 +46,25 @@ public ConnectorId getConnectorId()
     @Override
     public CompletableFuture<List<Split>> getNextBatch(int maxSize)
     {
-        return fetchSplits(bufferSize, maxSize);
+        checkArgument(maxSize > 0, "Cannot fetch a batch of zero size");
+        ConcurrentLinkedQueue<Split> result = new ConcurrentLinkedQueue<>();
+        CompletableFuture<?> future = fetchSplits(bufferSize, maxSize, result);
+        return future.thenApply(ignored -> ImmutableList.copyOf(result));
     }
 
-    private CompletableFuture<List<Split>> fetchSplits(int min, int max)
+    private CompletableFuture<?> fetchSplits(int min, int max, ConcurrentLinkedQueue<Split> output)
     {
         if (source.isFinished() || max == 0) {
```

Comment:
I think this would be simpler if you keep `min` and `max` constant throughout the recursion, and form the checks relative to `output.size()`